### PR TITLE
Imagej script

### PR DESCRIPTION
**Pull request by jburel**
This PR adds the ability to use the session ID and the selected objects in the UI in script 
to test:
 * Copy the imagej plugin into the plugins directory of Fiji
 * Connect to OMERO
 * Select an image. If you select any other object, only the object type should be printed out after connection
 * In Fiji go to ``File>Script ...``
 * Select ``Language > Groovy``
 * Run the following code snippet
```
// OMERO Dependencies
import omero.gateway.Gateway
import omero.gateway.LoginCredentials
import omero.gateway.SecurityContext
import omero.gateway.facility.BrowseFacility
import omero.gateway.facility.DataManagerFacility
import omero.gateway.model.DatasetData
import omero.log.SimpleLogger

import ome.formats.importer.ImportConfig
import ome.formats.importer.OMEROWrapper
import ome.formats.importer.ImportLibrary
import ome.formats.importer.ImportCandidates
import ome.formats.importer.cli.ErrorHandler
import ome.formats.importer.cli.LoggingImportMonitor

import loci.formats.in.DefaultMetadataOptions
import loci.formats.in.MetadataLevel

import ij.IJ

// Setup
// =====

// OMERO Server details
HOST = org.openmicroscopy.shoola.MainIJPlugin.getHost()
PORT = org.openmicroscopy.shoola.MainIJPlugin.getPort()
group_id = "-1"
USERNAME = org.openmicroscopy.shoola.MainIJPlugin.getSessionId()
PASSWORD = org.openmicroscopy.shoola.MainIJPlugin.getSessionId()

def connect_to_omero() {
    "Connect to OMERO"

    credentials = new LoginCredentials()
    credentials.getServer().setHostname(HOST)
    credentials.getServer().setPort(PORT)
    credentials.getUser().setUsername(USERNAME.trim())
    credentials.getUser().setPassword(PASSWORD.trim())
    simpleLogger = new SimpleLogger()
    gateway = new Gateway(simpleLogger)
    gateway.connect(credentials)
    return gateway

}

def open_image_plus(HOST, USERNAME, PASSWORD, PORT, group_id, image_id) {
    "Open the image using the Bio-Formats Importer"

    StringBuilder options = new StringBuilder()
    options.append("location=[OMERO] open=[omero:server=")
    options.append(HOST)
    options.append("\nuser=")
    options.append(USERNAME.trim())
    options.append("\nport=")
    options.append(PORT)
    options.append("\npass=")
    options.append(PASSWORD.trim())
    options.append("\ngroupID=")
    options.append(group_id)
    options.append("\niid=")
    options.append(image_id)
    options.append("] ")
    options.append("windowless=true view=Hyperstack ")
    IJ.runPlugIn("loci.plugins.LociImporter", options.toString())
}

// Connect to OMERO
println "connecting..."
gateway = connect_to_omero()


map = org.openmicroscopy.shoola.MainIJPlugin.getSelectedObjects();
for (entry in map) {
    println entry.key
    if (entry.key.contains("Image")) {
       println "opening Image..."
        image_id = entry.value[0]
        open_image_plus(HOST, USERNAME, PASSWORD, PORT, group_id, String.valueOf(image_id))
    }
}
println "Done"

```

cc @pwalczysko

### CODE REVIEW

**Comment by pwalczysko** [comment]
I have connected to merge-ci, not outreach. I think that is why the script did not work - could we maybe have an automatic prepopulation of the server address too ?

```
tarted New_.groovy at Thu Jun 27 10:46:46 BST 2019
omero.gateway.exception.DSOutOfServiceException: Login credentials not valid
	at omero.gateway.Gateway.connect(Gateway.java:267)
	at omero.gateway.Gateway$connect.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at Script1.connect_to_omero(Script1.groovy:42)
	at Script1.run(Script1.groovy:70)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:303)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:122)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.scijava.script.ScriptModule.run(ScriptModule.java:160)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:168)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:127)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:66)
	at org.scijava.thread.DefaultThreadService$3.call(DefaultThreadService.java:238)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
```

**Comment by pwalczysko** [comment]
I have edited the HOST line of the script, putting in the merge-ci address. But, I get

```
Started New_.groovy at Thu Jun 27 10:49:28 BST 2019
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:657)
	at java.util.ArrayList.get(ArrayList.java:433)
	at loci.ome.io.OmeroReader.initFile(OmeroReader.java:339)
	at loci.formats.FormatReader.setId(FormatReader.java:1397)
	at loci.plugins.in.ImportProcess.initializeFile(ImportProcess.java:499)
	at loci.plugins.in.ImportProcess.execute(ImportProcess.java:142)
	at loci.plugins.in.Importer.showDialogs(Importer.java:140)
	at loci.plugins.in.Importer.run(Importer.java:76)
	at loci.plugins.LociImporter.run(LociImporter.java:78)
	at ij.IJ.runUserPlugIn(IJ.java:228)
	at ij.IJ.runPlugIn(IJ.java:192)
	at ij.IJ.runPlugIn(IJ.java:181)
	at ij.IJ$runPlugIn.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133)
	at Script2.open_image_plus(Script2.groovy:65)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1218)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1027)
	at groovy.lang.DelegatingMetaClass.invokeMethod(DelegatingMetaClass.java:151)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine$2.invokeMethod(GroovyScriptEngine.java:283)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:69)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:52)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:154)
	at Script2.run(Script2.groovy:78)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:303)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:122)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.scijava.script.ScriptModule.run(ScriptModule.java:160)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:168)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:127)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:66)
	at org.scijava.thread.DefaultThreadService$3.call(DefaultThreadService.java:238)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by pwalczysko** [comment]
If I select a Project, then there is no error,

```
Started New_.groovy at Thu Jun 27 10:46:46 BST 2019
connecting...
Started New_.groovy at Thu Jun 27 10:49:28 BST 2019
connecting...
opening Image...
omero.gateway.model.ImageData
Started New_.groovy at Thu Jun 27 10:51:16 BST 2019
connecting...
opening Image...
omero.gateway.model.ProjectData
Done
```

but no image really opened, not sure what the expected behaviour is here.

**Comment by pwalczysko** [comment]
If I connect to outreach, then I have better experience. When one image is selected, I have

```
Started New_.groovy at Thu Jun 27 10:55:26 BST 2019
connecting...
opening Image...
omero.gateway.model.ImageData
Done
```
and one image really opened in Fiji.

If I select a Dataset, I get

```
Started New_.groovy at Thu Jun 27 10:57:01 BST 2019
connecting...
opening Image...
omero.gateway.model.DatasetData
Done
```
which is probably as expected too.

**Comment by pwalczysko** [comment]
On nightshade, it works fine just like on outreach. Tried also Plate, it reports

```
Started New_.groovy at Thu Jun 27 11:06:37 BST 2019
connecting...
opening Image...
omero.gateway.model.PlateData
Done
```

**Comment by pwalczysko** [comment]
1 RFE https://github.com/ome/omero-insight/pull/53#issuecomment-506276896

Also, maybe we should clarify why merge-ci does not work

**Comment by joshmoore** [comment]
> Also, maybe we should clarify why merge-ci does not work

One possibility is that an attempt is made to create an unencrypted connection.

**Comment by jburel** [comment]
@pwalczysko good point, host and port should be retrieved as well so you do not have to edit in the script

**Comment by pwalczysko** [comment]
The fix with the port and hostname works as expected. 
Also all the other functionality works as expected (except images on merge-ci see https://github.com/ome/omero-insight/pull/53#issuecomment-506277720 - I also tried to connect with the lock locked on insight splashscreen and then run the script, but that did not help)

If the merge-ci is explainable (cc @joshmoore ), then this is good to merge.

**Comment by pwalczysko** [comment]
Not sure why the travis is  still pending.

**Comment by jburel** [comment]
Synch issue between travis and github. This PR is not needed in the next version i.e. gateway version bump.

**Comment by jburel** [comment]
--exclude

File: src/main/java/org/openmicroscopy/shoola/env/data/DataServicesFactory.java [hunk 1]
```diff
@@ -482,13 +484,24 @@ public void connect(UserCredentials uc)
 		LoginCredentials cred = new LoginCredentials();
         cred.getUser().setUsername(uc.getUserName());
         cred.getUser().setPassword(uc.getPassword());
-        cred.getServer().setHostname( uc.getHostName());
```
Comment (dominikl) [inline src/main/java/org/openmicroscopy/shoola/env/data/DataServicesFactory.java]: Do you think you could remove this change? Maybe it wouldn't conflict with my commit then https://github.com/ome/omero-insight/pull/66/files#diff-d3089c1dff7a3f6c789980fcb8bcb74a (going to replace these lines anyway in that commit).

Comment (jburel) [reply src/main/java/org/openmicroscopy/shoola/env/data/DataServicesFactory.java]: I have excluding the PR for now. I will revisit when I am done with other tasks

**Comment by joshmoore** [comment]
@jburel : do you see trying to get this in for 5.6? (Are we bumping insight's version number with the 5.6 server?)

**Comment by jburel** [comment]
Closing this PR for now . This can come in a point release.

**Pull request closed without merging**